### PR TITLE
[openuserjs] Add OpenUserJS service badges

### LINKED ISSUE #5505: Badge request : openuserjs.org, greasyfork.org
:clipboard: **Description**

Services : openuserjs.org, greasyfork.org

Information : Install count, rating, version code

Examples : 

OpenUserJS
GreasyFork

:link: **Data**

* No public API
* No key needed

Data can be directly accessed from the HTML file.

:microphone: **Motivation**

This feature can help assert a certain "reputation" based on install count for public scripts.

More specifically, I want to be able to show off that people have installed my userscript.

### PR DESCRIPTION
**Pull request by DenverCoder1**
Fixes #5505 (when merged along with #8080 being merged)
Partially resolved - #829

API endpoint - [https://openuserjs.org/meta/username/scriptname.meta.json](https://openuserjs.org/meta/Marti/oujs_-_Meta_View.meta.json)

Documentation - https://openuserjs.org/about/Frequently-Asked-Questions#q-does-openuserjs-org-have-meta-

image

### CODE REVIEW

**Comment by DenverCoder1** [comment]
All tests are passing locally

image

File: services/openuserjs/openuserjs-license.service.js
Comment (Martii) [inline services/openuserjs/openuserjs-license.service.js:24]: You do realize this is probably incorrect. All Userscript key values are always LIFO. If you are on the site you may read about it specifically when adding a script with a primary license under dual or more licensing.

OUJS Admin

Comment (DenverCoder1) [reply services/openuserjs/openuserjs-license.service.js:24]: Thanks, do you have an example of what that might look like? I haven't seen any with multiple values.

I'm assuming it would be an array with each entry having a `value` key?

Comment (Martii) [reply services/openuserjs/openuserjs-license.service.js:24]: > do you have an example of what that might look like?

See https://openuserjs.org/about/Frequently-Asked-Questions#q-does-openuserjs-org-have-meta- and the Unit Test script mentioned and also the human readable *(text stream)* at https://openuserjs.org/user/add/scripts *(you need to be logged in to see it)*.

File: services/openuserjs/openuserjs-license.service.js
Comment (Martii) [inline services/openuserjs/openuserjs-license.service.js:25]: There is some "grey area" in the oldest scripts before I fixed this... the TOS default is `MIT`... also the SPDX codes have changed at least twice for GPL so that makes it rather difficult to track since we don't usually edit the scripts. It is enforced as a requirement for several years to have at least a single license *(FOSS specifically)*.

Comment (DenverCoder1) [reply services/openuserjs/openuserjs-license.service.js:25]: By this I assume you mean "license" always appears in the data, so there's no need to check if it's there or not?

Comment (Martii) [reply services/openuserjs/openuserjs-license.service.js:25]: For the last few years yes however oldest scripts no. See https://openuserjs.org/scripts/sizzle/Textarea_Keyboard_Shortcuts/source which is the first script. Look at the homepage too to see how we handle it as best as we can.

Comment (DenverCoder1) [reply services/openuserjs/openuserjs-license.service.js:25]: Thanks for your help!

File: services/openuserjs/openuserjs-version.service.js
Comment (Martii) [reply services/openuserjs/openuserjs-version.service.js:28]: > Are both spellings of license / licence possible to appear in the JSON?

What does https://github.com/OpenUserJS/OpenUserJS.org/blob/master/public/pegjs/blockUserScript.pegjs#L69-L93 say it will do when a key of that is found? i.e. don't make shields work more than it should and let us do what we do best for the meta script API. You can always try test scripts with values on production or even load up an instance in development. There's a reason why the Unit Test script exists too... utilize it to see how we do things

File: services/openuserjs/openuserjs-downloads.tester.js
Comment (Martii) [inline services/openuserjs/openuserjs-downloads.tester.js:10]: Okay... presuming that you are polling the site already in these lines *(based off the commit summary)* you might be asking for a headache with picking `NotAUser`. Someone could actually make this account and you might be introducing a random element in shields testing routines. Usually I do my test routines against my own account but that is up to Shields on whether or not they want to allow that. As long as you let me know you could use my username as I am relatively sure I'll never publish a "NotAScript" phantom script... or your can choose your own account... either way someone, in its current state, could easily make the tests fail and someone will be barking on Shields about it.

Comment (DenverCoder1) [reply services/openuserjs/openuserjs-downloads.tester.js:10]: Sounds good to me, I'd be fine with setting it to use my own account

File: services/openuserjs/openuserjs-downloads.service.js
Comment (DenverCoder1) [inline services/openuserjs/openuserjs-downloads.service.js:15]: @NatoBoram Are you fine with your script being used as an example on shields.io? If not, I can change it.

File: services/openuserjs/openuserjs-downloads.service.js
Comment (Martii) [inline services/openuserjs/openuserjs-downloads.service.js:24]: So for symmetry sake since you did the `at` prototype elsewhere for the UserScript block *(love when they sneak these methods in and I miss them btw)* do you think you ought to do it on _all occurrences_ in this PR that are applicable? It's not a huge deal breaker but might be nice to have.

Comment (DenverCoder1) [reply services/openuserjs/openuserjs-downloads.service.js]: As there's only 1 item as of now, they seem to be equivalent, although I can do that for consistency :+1:

File: services/openuserjs/openuserjs-downloads.tester.js
Comment (Martii) [reply services/openuserjs/openuserjs-downloads.tester.js]: > I suppose for a similar reason, the other tests should probably be changed to scripts that I know won't be changing randomly

Yeah I would do that as well since those too use `.get` *(which I still presume is actually polling OUJS)* NatoBoram is active but doesn't seem to be available at this time. If you are absolutely set in your way of showcasing his script... fork it on OUJS by going to his source code page and Submit Code as Fork and it will back-reference the original. You can steer others towards the root script in your Script Info.

I always do the Meta View Unit Test script in examples as I know unless something catastrophic occurs with Userscripts it should always be there... that is another choice.

File: services/openuserjs/openuserjs-downloads.tester.js
Comment (Martii) [inline services/openuserjs/openuserjs-downloads.tester.js:6]: ex. Here's a `.get` that isn't under Shields and/or your control.

File: services/openuserjs/openuserjs-issues.tester.js
Comment (Martii) [inline services/openuserjs/openuserjs-issues.tester.js:7]: ex. Here's a `.get` that isn't under Shields and/or your control.

File: services/openuserjs/openuserjs-license.tester.js
Comment (Martii) [inline services/openuserjs/openuserjs-license.tester.js:12]: This should be fine since I'm the Admin and co-owner of the site. This is my only glory I get these days :smile_cat: .

**Review by Martii — CHANGES_REQUESTED** [review]
> I suppose for a similar reason, the other tests should probably be changed to scripts that I know won't be changing randomly

File: services/openuserjs/openuserjs-downloads.service.js
Comment (NatoBoram) [reply services/openuserjs/openuserjs-downloads.service.js:15]: Oh, sure, have fun!

File: services/openuserjs/openuserjs-license.tester.js
Comment (Martii) [reply services/openuserjs/openuserjs-license.tester.js:12]: Tell you what... if you *(and Shields)* prefer I can create some Unit Tests *(dummy metadata blocks only)* under https://openuserjs.org/users/author so it will make the code more readable. That is also a reserved account and shouldn't be going anywhere. I can also create dummy issues for the counts.

Just give me some names *(or I can do it)* and initial values.

Comment (DenverCoder1) [reply services/openuserjs/openuserjs-license.tester.js:12]: I was thinking of making some fake scripts on my own account, but if you want to officially create some, that would work great as well.

For testing the metrics, it just needs to have a positive number (eg. non-zero number of installs, non-zero number of issues, etc.)

Comment (Martii) [reply services/openuserjs/openuserjs-license.tester.js:12]: > it just needs to have a positive number 

It's a full account, just no auth, so anyone can affect the values... .however I should create the base issues for test scripts with my Admin account so others above my ranking know why it exists. Anyone will be able to up/downrate it too but I might consider restricting that in the source... doesn't mean it goes away just not list-able at this point in time if too far down the rating.

Comment (Martii) [reply services/openuserjs/openuserjs-license.tester.js:12]: There you go...  https://openuserjs.org/users/author/scripts left the fork of Meta View as negative rating. If you want different numbers on install count tap the install it a few random times *(but don't go overboard)*. If you want more generic script names tell me. I left it so it can showcase what you have done. 

For `NotAScript` that is absent for that Shields test.

Comment (DenverCoder1) [reply services/openuserjs/openuserjs-license.tester.js:12]: Looks good. 

Maybe there could be one to replace the "no version found" unit test as well for consistency.

This is currently the only one I can't replace with the "author" account yet https://openuserjs.org/scripts/DenverCoder1/Example_no_version

Comment (Martii) [reply services/openuserjs/openuserjs-license.tester.js:12]: Forked and done.

**Comment by Martii** [comment]
@calebcartwright 

Up to this point I see no other issues from our stand-point. I don't fully understand every bit of the code for this but that's all I can analyze for now. Official :+1:  from us. :smile_cat: 

OUJS Admin / OUJS Co-Owner

File: services/openuserjs/openuserjs-license.service.js
Comment (Martii) [inline services/openuserjs/openuserjs-license.service.js:26]: @DenverCoder1 

One more quickie... sorry... if it is absent, say at https://openuserjs.org/scripts/JamesSwift/Newegg_ComboEgger , we apply `MIT` as the TOS default.... so not sure if you want to throw an actual error or put in a static `MIT`. Absentees no longer occur but older scripts weren't enforced.

Comment (DenverCoder1) [reply services/openuserjs/openuserjs-license.service.js]: Done :+1:

**Comment by calebcartwright** [comment]
Thanks to all involved, it's nice coming into a PR that's already had thorough discussion and review :smile: 

Couple questions before I dive in:

* Is there any public documentation for the endpoint? Doesn't need to be anything official, swagger page or even wiki/markdown content in a repo. We integrate with a large number of services, many of which we aren't familiar with ourselves, and it's helpful to know where to go for more info/bug reports/etc.
* Are there rate limits applied to this endpoint, and if so what are those current limits?

**Comment by DenverCoder1** [comment]
> Is there any public documentation for the endpoint? Doesn't need to be anything official, swagger page or even wiki/markdown content in a repo. We integrate with a large number of services, many of which we aren't familiar with ourselves, and it's helpful to know where to go for more info/bug reports/etc.

The existence of the page is documented at https://openuserjs.org/about/Frequently-Asked-Questions#q-does-openuserjs-org-have-meta- The `.meta.json` endpoint is being used.

> Are there rate limits applied to this endpoint, and if so what are those current limits?

@Martii may have a full answer. From what I've seen, usual requests to `.meta.json` don't seem to have a very strict ratelimit if any at all. But for 404 pages (eg. scripts that don't exist), it seems to get a 429 after about 3 requests within every few seconds.

**Comment by calebcartwright** [comment]
Thanks @DenverCoder1. Would be good to get a feel for what those are and to see what our options are. We do have some caching knobs we can tune but want to know what we're working with.

Also as an aside, you don't necessarily have to worry about updating the branch until we get close to being ready to merge. It's more of a final check we have in place to make sure there's no conflicting work and that the changes will be fine atop the trunk branch, but it's not necessary to add merge commits any time an unrelated PR gets merged.

**Comment by Martii** [comment]
> Are there rate limits applied to this endpoint, and if so what are those current limits?

Currently no... see: 

https://github.com/OpenUserJS/OpenUserJS.org/blob/a9cbf7d6670d5232bae44f284129a63d758210fb/routes.js#L148-L150

This could change if it is abused but so far knock on things. We're thinking about pruning `@icon` or "nubbing" it to our basic icons depending on how much bandwidth comes into play as the user base increases... but for now it's minimally managed.

**Comment by calebcartwright** [comment]
I appreciate you sharing relevant snippets from the codebase but I'm going to ask some follow up questions anyway to avoid incorrect assumptions about the source code, gain a better understanding of potential future plans, to account for anything else in the system architecture that could be at apply above/before the running app, etc.

the meta json endpoint is the only one we'll need

* is my understanding correct that this is not currently rate limited, and there's no explicit plans to change that? (understand you reserve the right to change it in the future regardless)
* how frequently does the meta.json file get updated? we rely on cache pretty heavily but have a standard set of ttl periods used by default across all badges for a given family (version vs. build status vs. etc.), but we can override the ttl for individual badges too.
* admittedly anecdotal, but the rate limiting that gets applied on 404 requests for the meta endpoint appear to get reset quite quickly, and, fortunately for us, doesn't seem to have any impact on a request for a valid/found script. Am I correct in assuming that's the behavior we should be expecting? It's important for us because we wouldn't want a handful of 404-type badge requests we receive to limit/lock out Shields.io and our ability to serve up these badges

**Comment by Martii** [comment]
> is my understanding correct that this is not currently rate limited, and there's no explicit plans to change that? (understand you reserve the right to change it in the future regardless)

Correct and what I've already stated.

> how frequently does the meta.json file get updated?

Any time the script gets updated which can be as little as never or as maximum as reposting allows *(hence the short period you mentioned next)*. We do however cache in a browser for a maximum of 7 days for HTTP 1 caching and immediately on eTag for HTTP 1.1 if changed.

> Am I correct in assuming that's the behavior we should be expecting?

Same as your first question bullet point answer here.

> It's important for us because we wouldn't want a handful of 404-type badge requests we receive to limit/lock out Shields.io and our ability to serve up these badges.

If you have an endpoint similar to GitHub's *(for synchronizing a repo here to us... mentioned at https://api.github.com/meta ... possibly much simpler than GH's of course)* we'd be more than happy to add Shield's server IP to a special list.

Also note other UserScript sites *(including GitHub)* has rate limiting so most of these questions probably should be asked/checked as well. :smile_cat:

**Comment by calebcartwright** [comment]
@Martii 

> Correct and what I've already stated.

I appreciate you taking the time to participate in the review and dialog here as we consider potentially adding a feature to our platform that would primarily be utilized by users of your platform.

However, please keep in mind that we integrate with and provide badges for hundreds of platforms/services, the overwhelming majority of which, including OpenUserJS, we do not use ourselves and have no familiarity with. You are obviously very familiar with your service, but the discussion here has been heavy on acronyms and tribal knowledge, and relative to the other services we integrate with, there's not a lot of API documentation through which we could edify ourselves.

We're still trying to play catch up, and clarifying questions are being asked because we don't feel we had sufficient clarity. Even if things looked and felt sufficiently clear to you, I'd still kindly ask that you refrain from the "already answered" types of comments and their associated implications.

> If you have an endpoint similar to GitHub's (for synchronizing a repo here to us... mentioned at https://api.github.com/meta ... possibly much simpler than GH's of course) we'd be more than happy to add Shield's server IP to a special list.

Understand the inclination but this won't be viable for us. Our runtime is dynamically scaled and ephemeral with dynamic IP addresses. As such when we engage with vendors/maintainers of platforms that have concerns around load, we typically look to adjust cache settings on our end and/or change to a model where the Shields requests are authenticated using an account the vendor/maintainers have granted a higher rate limit.

However, this doesn't sound like something that can/should be solved in the abstract, especially given your confirmation around the not found rate limiting behavior. If the traffic Shields sends to OpenUserJS starts to become a problem down the road, please don't hesitate to reach out and we'll work to find an appropriate solution (we send info in the request headers that could be used to filter our traffic from the rest).

> Also note other UserScript sites (including GitHub) has rate limiting so most of these questions probably should be asked/checked as well. smile_cat

Sorry, but I don't understand what you're saying at all. Are saying that we need to ask/investigate the questions I've asked you about OpenUserJS against all the other services we integrate with? If so, thanks, but we're all set. I ask these questions of you because these are the standard set of questions we consider when adding support for a new service/platform, and we're very, very familiar with GitHub's rate limits and how to work with them at mass scale.

**Comment by Martii** [comment]
@calebcartwright

> including OpenUserJS, we do not use ourselves and have no familiarity with

I am quite familiar with that and we've been using Shields for about 10 years and I've personally dealt with this repo before. :smile: 

>  I'd still kindly ask that you refrain from the "already answered" types of comments and their associated implications.

Apologies that your perception was skewed on this. Directness is our goal not beating around the bush.

> Our runtime is dynamically scaled and ephemeral with dynamic IP addresses.

So is GitHub but we'll just treat Shields in the same traffic pattern as a Guest then. No big deal. Thought I'd offer an improvement on Shields via our site, if it wasn't already available, since the retrieval on NPM is about 98% instead of the alluded 100%... but is mostly reliable compared to others.

> please don't hesitate to reach out and we'll work to find an appropriate solution

That's actually why we are here now to keep the bugs to a minimum for now. :smile_cat:  And of course... there's always the dynamic badges anyhow for Authors to utilize if things change outside of my current Active Maintainer scope.

> Are saying that we need to ask/investigate the questions I've asked you about OpenUserJS against all the other services we integrate with?

Recommending being impartial yes as to not be monopolistic in nature.

> If so, thanks, but we're all set. I ask these questions of you because these are the standard set of questions we consider when adding support for a new service/platform, and we're very, very familiar with GitHub's rate limits and how to work with them at mass scale.

There's more than just GitHub. We chose GitHub as a comment discussion point as it is familiar to all of use since we are all here for objectivity. You skipped this validation within the last three weeks. Just hoping for some objectivity from Shields instead of what our team has observed and perceived as a dual standard. Again we don't beat around the bush and we do appreciate your service as do others.

Anyhow.... if you feel the need to ask more questions I am available at your leisure but you are reminded that literal answer will be given at all times through my successful communications and lifetime accolades. @DenverCoder1 has plenty of accolades as well and this is the first we have enjoyed dealing with his organization/himself as he keeps his objectivity which is what our community strives for. :+1:

**Comment by calebcartwright** [comment]
@Martii - I was trying to give you a gentle nudge but allow me to be more direct this time. You may believe you're simply being blunt/literal, but your comments are coming across with an increasingly flippant tone. I need you to stop and ask you to be mindful of this caution should you choose to continue engaging in the conversation.

Your comments give the impression of at least being mildly irked that I posted a summarizing/clarifying comment, and then noted that impression. Perhaps I misunderstood, but that's how they come across, and when it comes to intra-human communications I think the phrase "perception is reality" holds a lot of water. I hope it was just a simple misunderstanding, but I also apologize if I've offended in any way. Again, I understand you feel you've been unequivocally clear in your brief commentary, but please remember that your perception doesn't inherently mean your shorthand is actually clear to everyone else. It wasn't, but we're in a better state now.

> Recommending being impartial yes as to not be monopolistic in nature.

> There's more than just GitHub. We chose GitHub as a comment discussion point as it is familiar to all of use since we are all here for objectivity. You skipped this validation within the last three weeks. Just hoping for some objectivity from Shields instead of what our team has observed and perceived as a dual standard. Again we don't beat around the bush and we do appreciate your service as do others.
>
> Anyhow.... if you feel the need to ask more questions I am available at your leisure but you are reminded that literal answers will be given at all times through my successful communications and lifetime accolades. @DenverCoder1 has plenty of accolades as well and this is the first we have enjoyed dealing with his organization/himself as he keeps his objectivity which is what our community strives for. +1

This is over the top, and unnecessary. This sounds as if you feel OpenUserJS is being unfairly scrutinized, which is absolutely not the case. I will make the assumption that you are referring to the recent GreasyFork badges and are trying to conflate that instance to paint a broader, and incorrect, set of characterizations. To the best of my knowledge, neither of these two services had public documentation around their respective rate limits, however, rate limits still became a discussion point here because they were directly observed by the contributor and you were present as a maintainer of that service.

There's a pretty standard set of criteria we consider when adding badges for a new service, and when applicable, that includes rate limiting on the upstream service too. Not every one of those considerations is a "blocker" per se, but are things we try to wrap our heads around for each service.

<hr />

I'm simply trying to ensure we have due diligence on our part to make sure we're operating as a good citizen with your platform and that we can provide a good badge experience to our users, nothing more and nothing less.

Also want to stress there's no hard feelings, and I hope we'll be able to get these badges incorporated. I just wanted to make sure we keep the conversation constructive, respectful, and in accordance with the the type of dialog we strive for here in our community.

**Comment by Martii** [comment]
> I was trying to give you a gentle nudge but allow me to be more direct this time.

See [the gentle nudge](https://github.com/badges/shields/pull/8081#issuecomment-1156877090) that one of our developers gave me to pass on gently to your team. So I'll reiterate their concern by re-quoting their concerns as posted by me:

> Recommending being impartial yes as to not be monopolistic in nature.

... in which I'm starting to see the point due to your inappropriate escalation of the matter *(mountain out of a molehill)*. I will and have given you appropriate respect and would hope to see that in your responses from now on instead of attacking the messenger in poor taste. :smile: 

> Perhaps I misunderstood, but that's how they come across

Text is always one of the worst ways to communicate unfortunately. As I mentioned your perception wasn't intended as a consequence and I was politely setting the record straight. If you would like even more direct... you are projecting some inadequacies that may have occurred to you, or this project, in the past. I'm sorry that you feel that way. Please don't project. 

On the matter of perception mentioning:

> the discussion here has been heavy on acronyms and tribal knowledge

... is confusing to say the least as I've done my personal best to cite just about everything and not use acronyms. One of our repos rules of thumb is not to use them whenver possible *(like lgtm or tldr)*. If it doesn't meet your standards I apologize... I'll attempt to walk you through it even more which is always part of life *(learning and teaching)*. Part of the communication here is not just DenverCoder1, you, and myself but all of the others that follow. So when I actually answer, and spend the time to cover all questions *(hopefully all)* by using quotations... please do not assume that it is being "irked" which is also unprofessional. I spent the time to answer all bullet points.  A lot of time was spent examining this PR including how Shields currently operates and breaking down your projects code to a point.

If you want even more direct:

> This is where I get fuzzy on the Shields side.

... means exactly that I'm not seeing any kind of DoS protection, at least in this PR, that sticks out in the code. This is something Shields should probably make more clear *(or at least cite)* and where I toss the proverbial ball to you *(since you are the current maintainer for this)*... instead didn't get much back. You should have examined the headers if I were to assume the role you think I am playing *(but definitely was not)* even though you said your project does.

> neither of these two services had public documentation around their respective rate limits,

They vary according to time and instances of DoS. Ours has been public for many years and mentioned directly by the contributor of this PR and was also mentioned by myself at https://github.com/badges/shields/pull/8081#discussion_r896481892 . Sorry you must have missed that. GF also has rate limiting and it's more discussed in their issues in their repo primarily and https://greasyfork.org/en/discussions/development with https://greasyfork.org/en/discussions/greasyfork . I asked DenverCoder to check something for them as I was unable to get to dev station to confirm myself as I am on GF as well but not actively. There's nothing "covert" about our documentation... it's all there under industry standard terminology for searching *(which I'll do when asked and have available time)*. If you want definitive you have the link again. :smile: I'll probably add the minimally managed to it next cycle if not sooner for your projects clarification.

> however, rate limits still became a discussion point here because they were directly observed by the contributor and you were present as a maintainer of that service.

Yes because we do our best to pay attention to all linkage even if it's another project or more. I saw a potential issue with this PR even though wasn't directly summoned here and it  helped my understanding on NatoBorams original request even better than the last time. In addition it gave some insights into the service badges vs dynamic badges.


So teaching is something that should be accepted and not interpreted as a Psychology exam of being "over the top" in which case you have already projected multiple times. I really don't have time to have philosophical debates with your trying to incorrectly analyze the situation. Don't read into things instead listen and be considerate like I have been. If you want "nasty" to justify your position we'll take things elsewhere... it's as simple as that. :smile: 

If you have any technical questions I'm still available. Ask away and I'll quote/answer away. Duplication happens a lot and once there is experience under ones belt projection ceases. Apologies for the novel but you escalated beyond the topic of the PR with the projections and I'm cutting it off.

**Comment by calebcartwright** [comment]
Minimizing the last comment as it's rather adversarial and there's no need to continue that thread which is increasingly off topic.

Thanks for your feedback on the submitted code, technical insight, and willingness to answer questions about your service. We'll be sure to reach out if there are any additional questions you can assist with!

File: services/openuserjs/openuserjs-license.service.js
Comment (calebcartwright) [inline services/openuserjs/openuserjs-license.service.js:24]: Perhaps this is something already well known to all users of the OpenUserJS platform regardless of how new they are to it, but I think it would be in our best interest to also mention this in our user facing documentation that's shown on the Shields.io page.

You can do this by adding a `documentation` key to the object in `examples` with whatever text you think will make the most sense, but if you choose to use the same wording as what's in this comment I'd suggest converting `TOS` to `terms of service` for explicitness

File: services/openuserjs/openuserjs-license.service.js
Comment (calebcartwright) [inline services/openuserjs/openuserjs-license.service.js:31]: Similar comment here as one I mentioned on another of your PRs, but this `transform` function has enough going on that I think unit tests would be helpful for us to have going forward. I can share some examples of how we do this if you aren't already familiar

File: services/openuserjs/openuserjs-version.service.js
Comment (calebcartwright) [inline services/openuserjs/openuserjs-version.service.js:28]: Same comment about schema compatibility (a missing `version` is handled, but not a `version: []`)

**Review by calebcartwright — COMMENTED** [review]
Thanks again @DenverCoder1! This is in a pretty good state.

I left a few inline items below on mostly small things, and noted another one I want to check on too.

File: services/openuserjs/openuserjs-downloads.service.js
Comment (calebcartwright) [reply services/openuserjs/openuserjs-downloads.service.js:6]: Yup, leave the route as-is for now and I'll follow up with a resolution one way or another. I believe using the download nomenclature in the route is intentional but it's been a while since we last discussed as time and my memory is a bit fuzzy on where we landed

File: services/openuserjs/openuserjs-issues.service.js
Comment (DenverCoder1) [reply services/openuserjs/openuserjs-issues.service.js:24]: It's the number of open issues for the script. It is typically used for bug reporting, you can see an example here:

https://openuserjs.org/scripts/laidbacktempo/EasyVideoDownload/issues/open

Closed issues are not included in the issue count.

**Comment by DenverCoder1** [comment]
It seems everything should be resolved now (besides waiting for the route name for installs)

File: services/openuserjs/openuserjs-downloads.service.js
Comment (calebcartwright) [reply services/openuserjs/openuserjs-downloads.service.js:6]: We'll likely continue discussing the specifics of this offline, but for now I think we've reached a sufficient consensus that it's fine to go with this route as is. We can always add redirectors (our mechanism for updating routes without impacting existing users) for these badges down the road if we end up deciding to pivot on the route conventions.

I'll try to make time over the weekend to do a final pass over the subsequent updates

Comment (calebcartwright) [reply services/openuserjs/openuserjs-downloads.service.js:6]: Actually, I see after reading through the previously linked issues that the maintainers of the upstream service still have some concerns, so we'll put this on hold for now until we figure out whether there's a viable path forward or whether we should just abandon this.

**Comment by calebcartwright** [comment]
@Martii

I was really hoping this matter was settled, but given subsequent discussion https://github.com/OpenUserJS/OpenUserJS.org/issues/1961, and particularly https://github.com/OpenUserJS/OpenUserJS.org/issues/1961#issuecomment-1158064065 onward, I see that it is not.

First and foremost, everyone that decides to participate in our community is inherently agreeing to our [Code of Conduct](https://github.com/badges/shields/blob/master/CODE_OF_CONDUCT.md), and that includes being subject to moderation from those of us that lead this community.

As stated in that document, we, the leaders of this community, are responsible for making the determination about what is, and what is not appropriate. Whether or not you like or agree with the rules of our community, or the decisions we make around appropriateness, does not change the fact that we have the final, and only say on this front within our community.

This all started because the way you had been communicating within our community on multiple occasions, particularly when responding to questions from both contributors and maintainers, was not really in accordance with the type of dialog we strive for here. It was still very mild at that point, and I assumed it was unintentional on your part, so I gave you what I believe was an in-kind, very mild moderation check in https://github.com/badges/shields/pull/8081#issuecomment-1175170525 in the hopes you'd have a typical "sorry, wasn't my intent" type of response. In retrospect I probably should have been more explicit about that motivation for that check, so apologies if that wasn't clear.

Unfortunately your reaction to that check, and especially the reaction to the subsequent moderation check, was simply inappropriate. I recognize that you disagree, that you think your behavior is fine, and that you did not appreciate those checks. However, your perception of the appropriateness of your behavior does not particularly matter in this context. You do not dictate the terms here and you do not get to determine what is and what is not appropriate in our community.

This is not up for debate; there is no need for you to defend or rationalize your prior responses, nor is there any utility in critiquing me or the project, so please do not do so again.

I have no intention of responding to the various name calling, misquotes, and insinuations from your previous comments above and in the linked issue in your project, however, I recognize that in some of those same comments you noted some continued technical concerns and I agree with you that we would need to discuss and come to an agreeable resolution before we could consider moving forward with this PR. I am willing to discuss those specifics here, provided you are willing and able to do so in a different manner, one that's in accordance with our community rules and goals, even if it requires agreeing to disagree about the prior discussion. (For what it's worth I'd also happy to answer questions over in your community if you'd prefer that, although it sounds like you preemptively banned me over there)

However, if you have decided that you would instead prefer that Shields not integrate with OpenUserJS, for any reason, we will absolutely abide by that, and abandon this PR. We could also consider an enhancement to our DynamicBadge feature to prevent users from leveraging that to produce OpenUserJS badges. I'm unclear on what if any relationship exists between OpenUserJS and GreasyFork, but the above can apply to GreasyFork as well if that's desirable/necessary.

Please let me know if you'd prefer that we abandon the OpenUserJS badges, or if you'd still like to see if there's a viable path forward, in which case I'll try to respond to some of your commentary around rate limits and dos.

**Pull request closed without merging**
